File: Dizzo-Studio-Mobile/android/app/src/main/kotlin/uz/dizzo/studio/TelegramLoginPlugin.kt
package uz.dizzo.studio

import android.app.Activity
import android.net.Uri
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.telegram.login.TelegramLogin

/**
 * Native "Log in with Telegram" (org.telegram:login-sdk) behind the
 * `uz.dizzo/telegram_login` channel. Dart side:
 * `lib/features/auth/data/telegram_native_login.dart`.
 *
 * - `init {clientId, redirectUri, scopes}`
 * - `login` → the id_token (String), null when cancelled/superseded, or an
 *   error (`not_configured`, `unavailable`, `failed`).
 * - `cancel` → completes a pending `login` with null.
 *
 * Telegram (or the Custom Tab fallback) returns through the App Link
 * `https://app{id}-login.tg.dev/tglogin`, which [TelegramLoginCallbackActivity]
 * receives and passes to [handleCallback].
 */
class TelegramLoginPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    private var channel: MethodChannel? = null
    private var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL).also { it.setMethodCallHandler(this) }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        finish(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "init" -> {
                val clientId = call.argument<String>("clientId").orEmpty()
                val redirectUri = call.argument<String>("redirectUri").orEmpty()
                val scopes = call.argument<List<String>>("scopes") ?: listOf("profile")
                if (clientId.isEmpty() || redirectUri.isEmpty()) {
                    result.error("not_configured", "clientId/redirectUri missing", null)
                    return
                }
                TelegramLogin.init(clientId = clientId, redirectUri = redirectUri, scopes = scopes)
                callbackHost = Uri.parse(redirectUri).host
                result.success(null)
            }
            "login" -> {
                val current = activity
                if (current == null) {
                    result.error("unavailable", "No activity", null)
                    return
                }
                if (callbackHost == null) {
                    result.error("not_configured", "init was not called", null)
                    return
                }
                // A newer attempt replaces an older one (the user tapped "open again").
                finish(null)
                pending = result
                try {
                    TelegramLogin.startLogin(current)
                } catch (e: Exception) {
                    Log.w(TAG, "startLogin failed", e)
                    fail("failed", e.message ?: "startLogin failed")
                }
            }
            "cancel" -> {
                finish(null)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    companion object {
        const val CHANNEL = "uz.dizzo/telegram_login"
        private const val TAG = "TelegramLogin"

        private var pending: MethodChannel.Result? = null
        private var callbackHost: String? = null

        private fun finish(idToken: String?) {
            val result = pending ?: return
            pending = null
            result.success(idToken)
        }

        private fun fail(code: String, message: String) {
            val result = pending ?: return
            pending = null
            result.error(code, message, null)
        }

        /** Returns true when [uri] was the Telegram login callback. */
        fun handleCallback(uri: Uri): Boolean {
            val host = callbackHost
            if (host == null || !uri.host.equals(host, ignoreCase = true)) return false
            if (pending == null) return true // app was restarted: nothing waits for it
            if (uri.getQueryParameter("error") == "access_denied") {
                finish(null)
                return true
            }
            try {
                TelegramLogin.handleLoginResponse(
                    uri,
                    onSuccess = { data -> finish(data.idToken) },
                    onError = { error ->
                        Log.w(TAG, "login failed: ${error.message}")
                        fail("failed", error.message)
                    },
                )
            } catch (e: Exception) {
                // No login session in the SDK (e.g. the process was killed meanwhile).
                Log.w(TAG, "handleLoginResponse failed", e)
                fail("failed", e.message ?: "No active login session")
            }
            return true
        }
    }
}

File: Dizzo-Studio-Mobile/ios/Runner/TelegramLoginPlugin.swift
import Flutter
import TelegramLogin
import UIKit

/// Native "Log in with Telegram" (the TelegramLogin Swift package) behind the
/// `uz.dizzo/telegram_login` channel. Dart side:
/// `lib/features/auth/data/telegram_native_login.dart`.
///
/// - `init {clientId, redirectUri, scopes}`
/// - `login` → the id_token (String), nil when cancelled/superseded, or an
///   error (`not_configured`, `failed`).
/// - `cancel` → completes a pending `login` with nil.
///
/// The SDK opens Telegram, or its own ASWebAuthenticationSession sheet when
/// Telegram isn't installed. Telegram returns through the universal link
/// `https://app{id}-login.tg.dev` (Associated Domains, Runner.entitlements),
/// which the scene delegate hooks below hand to `TelegramLogin.handle`.
final class TelegramLoginPlugin: NSObject, FlutterPlugin, FlutterSceneLifeCycleDelegate {
  static let channelName = "uz.dizzo/telegram_login"

  private var callbackHost: String?
  private var pending: FlutterResult?
  /// Bumped on every login, so a superseded SDK completion is ignored.
  private var attempt = 0

  static func register(with registrar: FlutterPluginRegistrar) {
    let instance = TelegramLoginPlugin()
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
    registrar.addSceneDelegate(instance)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
      let args = call.arguments as? [String: Any] ?? [:]
      let clientId = args["clientId"] as? String ?? ""
      let redirectUri = args["redirectUri"] as? String ?? ""
      let scopes = args["scopes"] as? [String] ?? ["openid", "profile"]
      guard !clientId.isEmpty, let host = URL(string: redirectUri)?.host else {
        result(FlutterError(code: "not_configured", message: "clientId/redirectUri missing", details: nil))
        return
      }
      // Channel calls arrive on the main thread.
      MainActor.assumeIsolated {
        TelegramLogin.configure(clientId: clientId, redirectUri: redirectUri, scopes: scopes)
      }
      callbackHost = host
      result(nil)
    case "login":
      guard callbackHost != nil else {
        result(FlutterError(code: "not_configured", message: "init was not called", details: nil))
        return
      }
      finish(nil)  // a newer attempt replaces an older one ("open again")
      attempt += 1
      let current = attempt
      pending = result
      MainActor.assumeIsolated {
        TelegramLogin.login { [weak self] outcome in
          DispatchQueue.main.async {
            self?.complete(outcome, attempt: current)
          }
        }
      }
    case "cancel":
      attempt += 1
      finish(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func complete(_ outcome: Result<LoginData, Error>, attempt current: Int) {
    guard current == attempt else { return }
    switch outcome {
    case .success(let data):
      finish(data.idToken)
    case .failure(let error as TelegramLoginError) where error == .cancelled:
      finish(nil)
    case .failure(let error):
      NSLog("Telegram login failed: \(error.localizedDescription)")
      fail(error.localizedDescription)
    }
  }

  private func finish(_ idToken: String?) {
    guard let result = pending else { return }
    pending = nil
    result(idToken)
  }

  private func fail(_ message: String) {
    guard let result = pending else { return }
    pending = nil
    result(FlutterError(code: "failed", message: message, details: nil))
  }

  /// True when `url` is the Telegram login callback.
  private func handleCallback(_ url: URL?) -> Bool {
    guard let url, let host = callbackHost,
      url.host?.caseInsensitiveCompare(host) == .orderedSame
    else { return false }
    Task { @MainActor in TelegramLogin.handle(url) }
    return true
  }

  // MARK: - Scene life cycle (the app uses UIScene / FlutterSceneDelegate)

  func scene(_ scene: UIScene, continue userActivity: NSUserActivity) -> Bool {
    guard userActivity.activityType == NSUserActivityTypeBrowsingWeb else { return false }
    return handleCallback(userActivity.webpageURL)
  }

  func scene(_ scene: UIScene, openURLContexts URLContexts: Set<UIOpenURLContext>) -> Bool {
    URLContexts.contains { handleCallback($0.url) }
  }

  // MARK: - App delegate (in case the scene delegate isn't used)

  func application(
    _ application: UIApplication,
    continue userActivity: NSUserActivity,
    restorationHandler: @escaping ([Any]) -> Void
  ) -> Bool {
    guard userActivity.activityType == NSUserActivityTypeBrowsingWeb else { return false }
    return handleCallback(userActivity.webpageURL)
  }
}
